### PR TITLE
feat(types): add `InferAttributes` utility type

### PR DESCRIPTION
**Pull request by ephys**
### Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

This PR introduces a few new utility types:
- `InferAttributes<M, opts>`: Used to drastically reduce the boilerplate needed to define Model Attributes in the class-based approach.
- `InferCreationAttributes<M, opts>`: Like `InferAttributes` but fields tagged with `CreationOptional<x>` are optional.
- `CreationOptional<x>`: A branded type used to mark an attribute as optional in Creation Attributes.
- `NonAttribute<x>`: A branded type used to omit a property from the attribute list.
- `Attributes<M>`: An alias for `M['_attributes']` for public use. Also forward compatible with planned v7 changes.
- `CreationAttributes<M>`: Same as `Attributes` but types that accept `undefined` are also `optional`.

To reduce risk of collision between `Model` and its subclasses, `Model#_attributes` and `Model#_creationAttributes` have been deprecated (I'd like to replace them with symbols in v7). Use `Attributes<Model>` and `CreationAttributes<Model>` instead.

In v7, we should replace these two with non-exported unique symbols.

### CODE REVIEW

File: docs/manual/other-topics/typescript.md
Comment (WikiRik) [inline docs/manual/other-topics/typescript.md:332]:
```suggestion
  // id can be undefined during creation when using `autoIncrement`
```

To make sure people know where it refers to

File: docs/manual/other-topics/typescript.md
Comment (WikiRik) [inline docs/manual/other-topics/typescript.md:338]:
```suggestion
  // createdAt can be undefined during creation
```

File: docs/manual/other-topics/typescript.md
Comment (WikiRik) [inline docs/manual/other-topics/typescript.md:369]:
```suggestion
  // createdAt can be undefined during creation
```

File: docs/manual/other-topics/typescript.md
Comment (WikiRik) [inline docs/manual/other-topics/typescript.md:381]:
```suggestion
  // updatedAt can be undefined during creation
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:2]: This note is only for the contributors of Sequelize right? Even though, we should put the complete path to the file

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:14]:
```suggestion
  // id can be undefined during creation when using `autoIncrement`
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:20]:
```suggestion
  // createdAt can be undefined during creation
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:22]:
```suggestion
  // updatedAt can be undefined during creation
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:51]:
```suggestion
  // createdAt can be undefined during creation
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:53]:
```suggestion
  // updatedAt can be undefined during creation
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:61]:
```suggestion
  // createdAt can be undefined during creation
```

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (WikiRik) [inline types/test/typescriptDocs/ModelInitWithAttributesOf.ts:63]:
```suggestion
  // updatedAt can be undefined during creation
```

**Comment by ephys** [comment]
No no, thank you for the review :) I did parts of it in a hurry to go eat and didn't notice most of the things you mentioned

File: docs/manual/other-topics/typescript.md
Comment (ephys) [reply docs/manual/other-topics/typescript.md]: It's meant for us yes. I've hidden it in a HTML comment so it's only visible in the raw md file. I've changed existing notes like this too.

File: types/test/typescriptDocs/ModelInitWithAttributesOf.ts
Comment (ephys) [reply types/test/typescriptDocs/ModelInitWithAttributesOf.ts]: Done, as well as the other files like this

**Comment by ephys** [comment]
The mapped types I'm using in this PR require TS >= 4.1. We'd need to drop support for TS <= 4.1 to merge this.

I'm on board with it. It's more than a year old now. The question is: do we want to consider that a semver major or minor?

**Comment by WikiRik** [comment]
> The mapped types I'm using in this PR require TS >= 4.1. We'd need to drop support for TS <= 4.1 to merge this.
> 
> I'm on board with it. It's more than a year old now. The question is: do we want to consider that a semver major or minor?

We had this discussion in another PR as well and I had difficulty with finding a good support plan for TypeScript versions. One that I did find was [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window) which tests on TypeScript versions that are less than 2 years old. But if you find other examples I would be fine with shortening the support window.

Since TypeScript releases come out approximately every three months and the minority of users use TypeScript, I would make this a semver minor change.

**Comment by ephys** [comment]
We can use `typesVersions` to expose different typing files based on TS version. I can try to do something with it.
Alternatively we drop support for versions `< 4.1` 8 months early (`4.0` having been released in august 2020)

**Comment by ephys** [comment]
Before we merge this, I'm also thinkings of ways to solve the `TCreationAttributes` part:

I have a solution that uses a type as branding to mark attributes as optional in TCreationAttributes:

image

It's a bit hacky but so far it works.

If we implement this as well as https://github.com/sequelize/meetings/issues/14, in v7 we should be able to simplify the Model definition to simply:

```
class User extends Model<User> {}
```

`Model` should be able to extract all the necessary information from the `User` class directly

File: docs/manual/other-topics/typescript.md
Comment (AllAwesome497) [inline docs/manual/other-topics/typescript.md:327]: Is this `declare` needed?

Comment (ephys) [reply docs/manual/other-topics/typescript.md:327]: Currently yes, due to https://sequelize.org/v6/manual/model-basics.html#caveat-with-public-class-fields

Also relevant is this https://github.com/sequelize/meetings/issues/16

**Review by AllAwesome497 — COMMENTED** [review]
Not an issue with this PR, but I think these associations are a bit too complex - its very easy to forget some.

File: docs/manual/other-topics/typescript.md
Comment (ephys) [reply docs/manual/other-topics/typescript.md:335]: It's a copy of the existing documentation but I agree we could simplify this to only the important bits. It's up to users to decide if they want it readonly.

If we agree on how to handle creation attributes with this method I think I'll just overhaul the whole documentation on TypeScript (see https://github.com/sequelize/sequelize/pull/13909#issuecomment-1005820419)

**Comment by ephys** [comment]
I've added a solution for creation attributes, but I went with a different direction than the branded type as it was causing issues.

Instead I went with a second option for `AttributesOf`: `optional`. This way it can be used for `TCreatrionAttributes` too:

```typescript
class User extends Model<
  AttributesOf<User, { omit: 'projects' }>,
  AttributesOf<User, { omit: 'projects', optional: 'id' }>
> {
  declare id: number;
  declare name: string;
  declare projects?: Project[];
}
```

Of course this double definition is due to backward compatibility requirements in v6.

In v7, I'll make it look like this instead:

```typescript
class User extends Model<User, { omit: 'projects', creationOptional: 'id' }> {
  declare id: number;
  declare name: string;
  declare projects?: Project[];
}
```

Unless we find a solution for branded types

**Comment by ephys** [comment]
Oh no, I found a way to make a branded type that works.

I'll revert some things.

image

**Comment by ephys** [comment]
I think this PR is almost ready:

- Added two utility types: `AttributesOf` & `CreationAttributesOf`
- Added two Branding types: `NonAttribute` & `CreationOptional`
- Overhauled the TypeScript documentation
- Had to mark all CreationAttributes that accept `undefined` as `optional` too, but I think that's a good change. :)
- Added lots of tests

Still needs to be done:

- Check that every instance of `_creationAttributes` is handled.
- Maybe replace `MakeUndefinedOptional<TModel['_creationAttributes']>` with a `GetCreationAttributes<TModel>` type to reduce duplication. (Note that it's different from `CreationAttributesOf`)
- Implement documentation updates from review

Going to put it back as draft until then.

Part of the Branding Type is a bit hacky. I've added lots of tests to make sure it doesn't negatively impact users. If anyone thinks of a scenario that would break don't hesitate to tell me.

File: docs/manual/other-topics/typescript.md
Comment (ephys) [reply docs/manual/other-topics/typescript.md]: Done :)

File: docs/manual/other-topics/typescript.md
Comment (ephys) [reply docs/manual/other-topics/typescript.md]: Added :)
Never realised there were that many. All the more reasons in favor of something like https://github.com/sequelize/meetings/issues/14

**Comment by ephys** [comment]
Ok now I think I'm fully done.

We just need to decide for TypeScript < 4.1.

I'm also still uncertain about using Branded Types. It feels fragile. But so far it works for all the different types I've included.

**Comment by ephys** [comment]
Well that's annoying, esbuild fails on numeric separators & optional catch bindings. I'm not sure why it didn't complain about it during the eslint PR. I'll revert these changes and we'll do them again once https://github.com/sequelize/sequelize/pull/13914 is finalized

**Comment by WikiRik** [comment]
esbuild or esdoc?

**Comment by ephys** [comment]
esdoc* :)

**Comment by ephys** [comment]
I think I figured what the issue was. It's not related to the new linting rules. It just did not like seeing HTML comments in a markdown file.

**Comment by ephys** [comment]
I think this is good to go now, just need a review 🙂

File: docs/manual/other-topics/typescript.md
Comment (WikiRik) [inline docs/manual/other-topics/typescript.md:47]: Maybe include which version of v6 adds this?

File: types/lib/associations/belongs-to-many.d.ts
Comment (WikiRik) [inline types/lib/associations/belongs-to-many.d.ts:3]: Maybe put it on a new line?

File: types/lib/associations/has-many.d.ts
Comment (WikiRik) [inline types/lib/associations/has-many.d.ts:3]: And put this on a new line as well?

File: types/lib/hooks.d.ts
Comment (WikiRik) [inline types/lib/hooks.d.ts:101]: Maybe add an expiring TODO for this?

Comment (WikiRik) [reply types/lib/hooks.d.ts:101]: Or do you think marking it as deprecated is enough for us to remember to remove it when we get closer to the v7 betas?

File: types/lib/sequelize.d.ts
Comment (WikiRik) [inline types/lib/sequelize.d.ts:22]: Maybe put these in three lines?

**Review by WikiRik — COMMENTED** [review]
Looks good. Few small remarks. I want to wait for the approval of @AllAwesome497 before we merge

File: types/lib/associations/has-many.d.ts
Comment (ephys) [reply types/lib/associations/has-many.d.ts]: I wish the linter would handle that, maybe at some point

File: types/lib/hooks.d.ts
Comment (ephys) [reply types/lib/hooks.d.ts:101]: Done, added a TODO (although it won't be triggered because /types is ignored by eslint)

File: docs/manual/other-topics/typescript.md
Comment (ephys) [reply docs/manual/other-topics/typescript.md:47]: I don't know which version of v6 it is yet but we can do this just before a merge :)

**Review by AllAwesome497 — APPROVED** [review]
looks fine to me

File: types/lib/hooks.d.ts
Comment (ephys) [reply types/lib/hooks.d.ts:104]: I can't yet, it would break user code, but I can in v7

**Comment by ephys** [comment]
@WikiRik small documentation update: I've added the v6 expected release version, fixed the documentation using the old utility types, and documented `Attributes<Model>`, `CreationAttributes<Model>`, and `ModelStatic<Model>`

File: types/lib/hooks.d.ts
Comment (WikiRik) [reply types/lib/hooks.d.ts:104]: Maybe add that to the TODO as well?

Comment (ephys) [reply types/lib/hooks.d.ts:104]: Now that I think of it, if it's `private`, `Attribute<M>` will not be able to access it. It's going to have to be a public internal Symbol I think

**Pull request merged**
